Wgpu Backend

KimHenrikOtte opened the pull request.

EricLBuehler commented:
@KimHenrikOtte thanks for the PR. This is super exciting, please let me know when you are ready for review.

nosnakeob committed:
Fixed Quantized Qwen3 Model (#2951)

* optimize KV cache to reduce GPU memory usage

* revert to using candle_nn::kv_cache::KvCache with initial capacity of 512

b0r3k committed:
Make tensor contiguous before the repeat_kv calls to avoid strided copies (#2953)

janimo committed:
Olmo 2 model (#2954)

* OLMo 2 model

* Update olmo-2 to example

* Clippy fix.

---------

Co-authored-by: laurent <laurent.mazare@gmail.com>

maximizemaxwell committed:
Fix docs quantized qwen3 (#2955)

* fixed docs quantized-qwen3 README

* fixed docs quantized-qwen2-instruct README

LaurentMazare committed:
Fixes for clippy 1.87. (#2956)

LaurentMazare committed:
Proper support for phi-4 (#2960)

* Add phi-4 support.

* Long-rope support.

* Get clippy to be happy.:

LaurentMazare committed:
Use a tanh activation in the xlm-roberta classification head. (#2968)

klion26 committed:
(hotfix) fix the doc test for indexer (#2970)

jpe90 committed:
Add fine-tuned text classifier to xlm roberta example (#2969)

ssfdust committed:
feat: enhance linear algebra operations (#2972)

- Add `dot()` for vector/matrix products
- Implement the `Frobenius` norm
- Add `mv()` for matrix-vector multiply

greenrazer committed:
candle-onnx: Implement Trilu and ScatterND ops (#2952)

* onnx attention

* setup an example, adding and fixing onnx ops bit by bit

* model working, output is garbage data

* trilu working

* close but not quite, Issues still with scatterND

* closer but the outputs are still slightly wrong

* added tests for trilu and scatterND

* lint

* readme

* clippy

* removed unnessisary comments

* changed device selection, took hyperparameters from model config

greenrazer committed:
Add Qwen3 MoE (#2934)

* qwen-moe rebase

* lint

* fixed rebase error

* swapped normal MoE model with CausalMoE Model in example, and swapped the tie word embeddings if statement

* updated readme

akshayballal95 committed:
Fix cuda memory error for Qwen3 non-quantized (#2987)

* Update KvCache initialization in Qwen3 model to use a fixed max position embedding value of 512

* add doc

BrunoSienkiewicz committed:
candle-onnx: Implement RNN operator (#2964)

* add: wip RNN parameters

* fix: corrected access to tensor dim in rnn

* add: rnn function call

* merged files

* added parameter parsing

* update: rnn parameter parsing

* remove: ONNX descriptions

* update: implemented basic operations

* update: removed comment

* add: RNN test

* update: prepared test values

* fix: operations on tensors

* update: passing tests

* add: test gen script

* changed error message

---------

authored-by: misadowsk <michalsad.protondynamic@gmail.com>

omahs committed:
Fix typos (#2958)

Michall00 committed:
candle-onnx: Implement Hard Swish operator (#2980)

* feat: added Elu operator

* feat: added hard swish

* added more tests for hard swish

* clened up

---------

authored-by: misadowsk <michalsad.protondynamic@gmail.com>

greenrazer committed:
fixed errors with hardswish merge (#3006)

greenrazer committed:
Fixed Failing CI (#3007)

* fixed clippy and new rust fmt

* po3 clippy changes

* updated workflow for PYO3

* tell maturin that version will be determined dynamically

* changed names of maturin workflow to avoid naming conflicts when uploading wheels

* switch to macos-13 bc ring crypto dependency not working on arm based macs

* back to failing macOS-latest since we should be compiling against arm mac, and this error isn't our fault'

zackangelo committed:
Qwen3: fix quality loss due to rope freq precision (#3005)

* qwen3 bugfix: compute rope freqs in f32

* qwen example: run model in bf16 on metal

greenrazer committed:
fixed ring mac error (#3008)

guoqingbao committed:
Support new arch of GLM4 models (#2991)

* Support new arch of GLM4 models

* Clippy fix & update ReadMe

* Integrate old and new GLM4 into one example & fix eos and chat template bugs for old GLM4

* Remove either crate usage

* clippy

---------

Co-authored-by: keighbee <kb@huggingface.co>

Michall00 committed:
candle-onnx: Implement One Hot operator (#2979)

* feat: added Elu operator

* feat: added implementation of onehot

* added test for one hot

* feat: add handling negative indicies value in OneHot operator

* lint

* lint

---------

Co-authored-by: misadowsk <michalsad.protondynamic@gmail.com>
Co-authored-by: keighbee <kb@huggingface.co>

xavierforge committed:
fix (candle-datasets): re-export FileReader and simplify from_hub iterator logic

- Re-exported the FileReader trait from the parquet crate so users can call methods
  like .metadata(), .num_row_groups(), etc., without needing to depend on parquet directly.
- Simplified the from_hub iterator logic by replacing filter_map<Option<Result<_>>> with
  a clean filter + map + collect<Result<_, _>>() chain.
- Added doc comments and example usage to clarify trait import and improve API ergonomics.

slckl committed:
candle-datasets: add fashion-mnist (#3021)

Michall00 committed:
candle-onnx: Implement Selu operator (#2978)

* feat: added Elu operator

* feat: added selu implementation in candle-nn

* feat: added selu onnx operator implementation

* test: added test for selu onnx operator

* test: added more tests for selu

* deleted elu

* tests: added test based on onnx specification

* lint

---------

Co-authored-by: misadowsk <michalsad.protondynamic@gmail.com>
Co-authored-by: keighbee <kb@huggingface.co>

piedshag committed:
fix: DAC model prefix (#3020)

EricLBuehler committed:
*Major T/s improvement* Use the Metal qmatmul MM kernels  (#2615)

* Add GGUF BF16 support (#17)

* Add GGUF bf16 type support

* Add non avx impl for vec_dot_bf16

* Fix from_u32

* Fix loading

* Fix dequant of bf16

* Update kernels for metal bf16 (#19)

* Update kernels for metal bf16

* Fix typo

* Check if have bfloat

* Sync ggml metal kernels (#33)

* Metal qmatmul mat-mat product (#39)

* Test passes

* All tests pass

* Now all the tests really pass

* Try out always using mm

* Mirror llama.cpp metric

* Mirror llama.cpp metric

* Update test

* Update test

* fixed merge error

---------

Co-authored-by: keighbee <kb@huggingface.co>

strickvl committed:
Fix discord badge (#3033)

EthanAlmloff committed:
Added GradStore::insert_id(id, grad)

jncraton committed:
Support building on CPUs with AVX but not AVX2 (#3040)

This change corrects an issue causing AVX2 intrinsics to be called on
CPUs that do not support them. These intrinsics were conditionally
compiled in behind an `avx` feature gate. This change correctly uses
them only if `avx2` is available. This change should have no impact
on modern CPUs, but it allows older CPUs to work properly using the
unoptimized code path.

jorge-menjivar committed:
[FEAT] Voxtral Support (#3036)

* feat: implement some configs in voxtral

* fix: fixed imports, implement more func

* feat: implemented full version, need fixes

* fix: fixed some compile errors

* feat: add initial examples

* fix: fixed voxtral.rs

* fix: fixed compile errors in examples

* fix: fixed compile errors

* fix: update model integration

* First working example

* Remove unused melfilters code

* Remove unused code

* Reuse whisper's pcm_decode

* Simplify generation function

* Remove unnecessary post-process fun

* Reuse snac's resample

* Apply clippy suggestions

* Remove unused filters

* Improve example

* Update tekken-rs

* Clippy fixes

---------

Co-authored-by: Max <naturale@hufs.ac.kr>

greenrazer committed:
ignored url that was interpreted as a secret by trufflehog (#3046)

zackangelo committed:
fp8 support (#2989)

* fp8 support

* use float8 crate with cudarc 16.x, fix errors

* fix Tensor::ones for fp8

* fp8: fix failing tests

* more fp8

* add fp8 where bf16 is in tests

* skip fp8 testing on metal

* fixed onnx eval match statements that didn't have full coverage

* Unused import backend::BackendDevice

* kernels: fix cuda_arch guards for fp8 ops

---------

Co-authored-by: keighbee <kb@huggingface.co>

chadvoegele committed:
Load safetensors i8 (#3042)

guoqingbao committed:
Fix sort kernel launch bug when nrows exceed gridDim.y limit (65535) (#3050)

greenrazer committed:
clippy fixes (#3053)

rsb-tbg committed:
Add timestamp rules and constraints to decoder in Whisper example (#3054)

* Apply timestamp rules in whisper decoder and add support for maximum initial timestamp index

* Optimize mask generation in decoder by pre-allocating a reusable buffer

* Refactor timestamp probability calculations in decoder to use log-softmax for numerical stability

lucky-bai committed:
Fix wasm build by enabling getrandom wasm_js backend (#3055)

rgbkrk committed:
pick seed <= u32::MAX when using metal (#3045)

davenpi committed:
Fix broken slice_scatter example in basics.rs

- Change tensor b from [1,2] row vector to [2,1] column vector
- Fix assertion to match expected result after column replacement
- Resolves shape mismatch error that prevented example from running

davenpi committed:
Run cargo fmt on basics.rs

ivarflakstad committed:
Metal device.set_seed full u64 support (#3067)

* Add simple atomics to ulong via atomic_uintx2 struct

* Remove u32::max restriction from metal device.set_seed

ivarflakstad committed:
disable affine fp8 bench on metal as it is not supported yet (#3065)

ivarflakstad committed:
Bench using chosen device only (#3066)

ivarflakstad committed:
Fixes metal randn determinism. Ensure we use the 2 atomic_uints buffer correctly (#3069)

joeldsouzax committed:
build: Make build.rs candle-kernels compatible with Nix and sandboxed builds (#3059)

* Use OUT_DIR for generated PTX bindings

* fix: fixed the out_dir cargo problem in examples

* fix: added imports in build.rs

ivarflakstad committed:
[Metal] update to objc2_metal (#3064)

* Initial metal-rs -> objc2-metal conversion

* Using objc2_metal bindings in metal kernels

* Use objc2_metal for mlx kernels

* Use objc2_metal for tests

* Use objc2_metal for metal benchmarks

* tidy

* Remove AllocationError. Use existing FailedToCreateResource

* All candle-metal-kernels tests passing

* Fix set_threadgroup_memory_length, fmt

* Update cargo tomls with objc2 libs

* Update candle-core metal usage

* impl Send/Sync for metal Device and Library structs

* tidy up imports

---------

Co-authored-by: Kyle Birnbaum <kb@huggingface.co>

EricLBuehler committed:
Fused CPU attention kernels (~4x performance increase) (#2973)

* Add cpu flash attention

* Add test

* Format

* Fix docs shape

szepeviktor committed:
Fix typos

ivarflakstad committed:
Merge pull request #3072 from szepeviktor/typos

greenrazer committed:
Fix iOS app store validation issues (#3071)

* put ug cuda behind cuda flag

* revert to ug 0.0.2 when on ios

* if only use ug if target_os is not ios added to wasm check already there

ivarflakstad committed:
Merge pull request #3038 from NoodlesOfWrath/gradstore_insert_id

zhanluxianshen committed:
clean candle-core typos.

Signed-off-by: zhanluxianshen <zhanluxianshen@163.com>

ivarflakstad committed:
Ensure metal tensors are send/sync via thread isolated command buffer map

jhqxxx committed:
Update kv_cache.rs (#3035)

ivarflakstad committed:
[Metal] Refactor (#3070)

* Initial metal-rs -> objc2-metal conversion

* Using objc2_metal bindings in metal kernels

* Use objc2_metal for mlx kernels

* Use objc2_metal for tests

* Use objc2_metal for metal benchmarks

* tidy

* Remove AllocationError. Use existing FailedToCreateResource

* All candle-metal-kernels tests passing

* Fix set_threadgroup_memory_length, fmt

* Update cargo tomls with objc2 libs

* Update candle-core metal usage

* impl Send/Sync for metal Device and Library structs

* tidy up imports

* Move .metal files to src/kernels/

* Begin refactor of candle-metal-kernels

* Refactor candle-core metal usage

* Delete old tmp folder

* Extract library to its own file

* Tidy up imports

* Refactor metal buffer concepts

* Refactor metal commandbuffer

* Refactor metal blit and compute command encoders

* Refactor metal compute pipeline

* Refactor metal commands struct

* Refactor metal Kernels

* Refactor kernel Source

* Extract MetalKernelError to err file

* Rename kernels/ -> metal_src/

* Add kernels folder for specific kernel call impls

* Move unary impls into kernels/unary.rs

* Move binary impls into kernels/binary.rs

* Move ConstantValues impl to metal::library

* Move sdpa impls into kernels::sdpa

* Move quantized impls into kernels::quantized

* Move cast impls into kernels::cast

* Move reduce impls into kernels::reduce

Technically not all of these are reduce ops.
That will have to wait for another day.

* Move copy into kernels::cast. Simplify imports

* Move affine impls into kernels::affine

* Move ternary impls into kernels::ternary

* Move indexing impls into kernels::indexing

* Simplify imports

* Move random impls into kernels::random

* Move conv impls into kernels::convolution

Again several impls are not specifically convolution.
Another day.

* Move fill impl into kernels::fill

ivarflakstad committed:
Merge pull request #3077 from zhanluxianshen/typo-candle-core

clean candle-core typos.

ivarflakstad committed:
Fix metal exports (#3081)

ivarflakstad committed:
Merge branch 'main' into metal-tensor-fix-send-sync

ivarflakstad committed:
Merge pull request #3079 from huggingface/metal-tensor-fix-send-sync

[Metal] Ensure tensors are send/sync

ivarflakstad committed:
Merge pull request #3062 from davenpi/fix/core-basics-example

Fix broken slice_scatter example in basics.rs

jfernandez committed:
Add CUDA 13 support (#3078)

Update cudarc to v0.17.3 which has support for CUDA 13.

ivarflakstad committed:
Fix indentation

ivarflakstad committed:
Fix indentation (ok but for real)

ivarflakstad committed:
Merge pull request #2937 from ljt019/fix-phi3-kv-cache-reset

fix kv cache issue with quantized_phi3 implementation

oa-root committed:
add impl for mish activation function (#3051)

grahamking committed:
Upgrade ug dep for CUDA 13 support

ug 0.5.0 pulls in cudarc 0.17.3, which supports CUDA 13

Signed-off-by: Graham King <grahamk@nvidia.com>

ivarflakstad committed:
Merge pull request #3089 from grahamking/main

ivarflakstad committed:
Various minor improvements, some suggested by clippy

ivarflakstad committed:
Merge pull request #3023 from xavierforge/bug/metadata-method-not-found

ivarflakstad committed:
Add command buffer thread map. Remove unecessary failure points

ivarflakstad committed:
Merge pull request #3092 from huggingface/metal-clippy-fixes

[Metal] various minor improvements

ivarflakstad committed:
Merge branch 'main' into improve-metal-command-buffer-map

ivarflakstad committed:
Fix multiple clippy warnings (#3101)

* is_multiple_of clippy fixes

* candle-core clippy fixes

* candle-nn clippy fixes

* candle-transformers clippy fixes

* candle-wasm-examples clippy fixes

* Fix quantized avx clippy

* Fix candle-transformers clippy

ivarflakstad committed:
Merge branch 'main' into improve-metal-command-buffer-map

ivarflakstad committed:
Wait until completed on command buffer status: scheduled as well

ivarflakstad committed:
Add metal conv for more dtypes

ivarflakstad committed:
Allow based to run with bf16 on metal

ivarflakstad committed:
Add backtracing to metal kernel errors for clarity

ivarflakstad committed:
Merge pull request #3103 from huggingface/metal-fix-conv

[Metal] Add conv ops for more dtypes

slckl committed:
bump gemm dependency to 0.18.2 to match ug

ivarflakstad committed:
Merge pull request #3100 from huggingface/improve-metal-command-buffer-map

[Metal] Improve command buffer map

ivarflakstad committed:
Merge pull request #3105 from slckl/gemm-bump

bump gemm dependency to 0.18.2 to match ug

ivarflakstad committed:
[Metal] Buffer improvements (#3093)

* Improve metal buffer usage. Add gather u8 kernels

* Use same metal resource options for all buffers. Fix copy call in allocate_buffer_with_data

* Make allocate_buffer pub. remove new_buffer_managed

* Remove allocate_buffer_with_data for now (needs sync)

ivarflakstad committed:
Various quantization improvements. Direct copy. Verified block sizes. (#3097)

* Add direct copy for floats in quantization

* Calling q matmul directly gives slightly different results, but within ggml error leniency

* fix quantized_mm test. Flip a and b input to matmul

* Add compile time verification of block sizes being equal to vec dot type block sizes

* Since we have verified that the block sizes are equal we can simplify qmatmul

* clippy

* Improved direct copy. Add comment to debug assert

* Add more info to quantized matmul test failures

* Disable quantized_mm for bf16 temporarily

slckl committed:
make cuda benches run again (#3111)

ivarflakstad committed:
Capture command buffer errors if they exist (#3106)

ivarflakstad committed:
[Metal] Improve wait_for_completed command buffers locking (#3107)

ivarflakstad committed:
Skip unsupported quantized matmul tests for metal (#3115)

ivarflakstad committed:
Fix beit on metal by adding additional affine implementations (#3116)

matthewhaynesonline committed:
Rough example of inlining model files into binary (#3104)

ivarflakstad committed:
Where cond get_strided_index conditionally based on function constants (#1616)

hoodiecollin committed:
feat(stable-diffusion): add build_unet_sharded method  (#3118)

* feat(stable-diffusion): add build_unet_sharded method for multi-file weight loading

Add build_unet_sharded method to StableDiffusionConfig to support loading
UNet models from multiple sharded safetensors files. This is necessary for
large models that are distributed across multiple files to avoid memory
constraints during model storage and distribution.

The new method accepts a slice of file paths instead of a single directory,
allowing direct specification of shard files. It uses mmaped_safetensors
for efficient memory usage when loading from multiple weight files.

Key differences from build_unet:
- Accepts &[P] where P: AsRef<Path> for multiple file paths
- Uses from_mmaped_safetensors with file list instead of directory
- Maintains same signature for in_channels, use_flash_attn, and dtype

Use case: Loading models like FLUX or large SD3.5 variants that are
distributed as model-00001-of-00003.safetensors, model-00002-of-00003.safetensors, etc.

Example usage:
    "model-00001-of-00002.safetensors",
    "model-00002-of-00002.safetensors",
];
let unet = config.build_unet_sharded(&unet_files, &device, 4, false, DType::F32)?;

* Update candle-transformers/src/models/stable_diffusion/mod.rs

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

ivarflakstad committed:
Fix metal get_function error (#3114)

ivarflakstad committed:
Quantization use debug_assert in hot paths (#3109)

* Add direct copy for floats in quantization

* Calling q matmul directly gives slightly different results, but within ggml error leniency

* fix quantized_mm test. Flip a and b input to matmul

* Add compile time verification of block sizes being equal to vec dot type block sizes

* Since we have verified that the block sizes are equal we can simplify qmatmul

* Improved direct copy. Add comment to debug assert

* Add more info to quantized matmul test failures

* Use debug_assert instead of if-else + bail! in hot paths

* quantized avx debug_asserts

* quantized simd128 debug_asserts

atilag committed:
Adding inference for GraniteMoeHybrid models from IBM (#3117)

LLukas22 committed:
Fix failing `cuda` build (#3121)

xavierforge committed:
feat: add metal_if_available method for graceful Metal fallback (#3041)

ivarflakstad committed:
Fix metal quantized to_float calls (#3123)

slckl committed:
Add more conv2d bench cases to candle-nn benches (#3131)

ivarflakstad committed:
Fix single file binary builder to only run when env var is set (#3126)

whitebox2 committed:
Update modernbert.rs (#3010)

ivarflakstad committed:
Update dependencies (#3135)

* Update candle metal dependencies

* Update benches black_box usage

* Update candle onnx dependencies

* Update tokio in examples/book

kshitijl committed:
Ensure output of Transpose is contiguous to prevent downstream MatMul from crashing (#3088)

kshitijl committed:
In the BERT example: apply the attention mask from tokenization during pooling (#3085)

ivarflakstad committed:
Skip q8k and q8_1 tests on cuda (#3140)

EricLBuehler committed:
Implement qwen3 vl

EricLBuehler committed:
Clippy

ivarflakstad committed:
Bump candle version to 0.9.2-alpha.1 (#3146)

EricLBuehler committed:
Remove unused

EricLBuehler committed:
Add Qwen 3 VL to candle-transformers

Add Qwen 3 VL to candle-transformers

ivarflakstad committed:
Improve candle example buildtime downloader (#3147)

* No longer dependant on separate crate and feature flag. Use compile time env vars instead

Co-Authored-By: Matthew Haynes <70829360+matthewhaynesonline@users.noreply.github.com>

* Add feature flag to single binary example so clippy does not try (and fail) to build it

---------

Co-authored-by: Matthew Haynes <70829360+matthewhaynesonline@users.noreply.github.com>

slckl committed:
CPU Conv2d: separate module, tiled im2col, specialization (#3136)

slckl committed:
rust-ci: add --benches to clippy, fix warnings (#3148)

slckl committed:
candle-core: add `broadcast_add` benches (#3149)

neksodebe committed:
fix: build errors for compute cap 7.5 (#3142)

* fix: remove hmax_nan and hmin_nan definitions

these conflict with cuda_fp16.hpp of CUDA 13.0

* fix(compatibility.cuh): uncomment code, lower compute cap threshold to <7.5

this makes it so that __hmax_nan and __hmin_nan are only defined for compute caps lower than 7.5, which fixes build issues at that compute cap.

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

matthewhaynesonline committed:
Update cargo build instructions to use double colon syntax (#3132)

* Update cargo build instructions syntax to use double colon syntax

Follow up of https://github.com/huggingface/candle/pull/3104#discussion_r2409288382

* Change cargo::info statements to cargo::warning

See: https://github.com/huggingface/candle/pull/3132#issuecomment-3444034565

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

EricLBuehler committed:
Add flash attn v3: `candle-flash-attn-v3` (#3152)

EricLBuehler committed:
Add nccl feature to candle-core (#3155)

ivarflakstad committed:
clippy default fixes (#3160)

whitebox2 committed:
Fix Metal matmul failure in `ModernBertHead::forward` by ensuring contiguous input  (#3139)

EricLBuehler committed:
Add varbuilder get_unchecked methods (#3157)

* Add varbuilder get_unchecked methods

* Add set_device and set_dtype

EricLBuehler committed:
Add unsafe from_storage apis (#3156)

Commit:
Merge remote-tracking branch 'origin/main' into wgpu_cleanup

Commit:
fixed merge errors

Commit:
fixed tests

Commit:
fixed clippy warnings

Commit:
updated dependencies

ivarflakstad committed:
[Metal] Ensure metal backend is send/sync via status semaphore (#3164)

* Add command status semaphore used to ensure metal backend is send/sync

* Update metal backend to use encoders directly instead of command buffer for send/sync correctness

* Update metal candle-nn ops to use encoders directly instead of command buffer for send/sync correctness

* Clippy

vrdn-23 committed:
Add sqrt2 as constant for gelu_erf and use `libm` erf (#3168)

* Add sqrt2 as constant for gelu_erf

* fix formatting

* Use a better erf function

slckl committed:
candle-core: strided-index inline next + size_hint + exact size iterator (#3169)

pcuenca committed:
Fix DINOv2 no-interpolation shortcut (#3172)

ivarflakstad committed:
Use Tensor::argmax instead of manual cpu impl (#3173)

ivarflakstad committed:
Fix argmax. Higher index should also be taken into account (#3179)

anonenity committed:
Add command buffer pool for improved multi-threaded Metal performance (#3175)

* add initial pool implementation

* update implementation to fix breaking tests

* add pool based tests to main test suite

* fix ordering types to avoid race conditions

* add in flight processing

* improve error handling and add wait flush test

* ensure flush state is returned from entry

* rename vars for clarity

* address pr comments

* update error mapping

* update to select entry with max compute count

* update tests and set default pool size

DrJesseGlass committed:
feat(candle-nn) ConcatKvCache for 2-5x GPU speedup on autoregressive generation (#3143)

* add concat cache; use in qwen3

* update tradeoff desc; resolve unused var warning in concatKV test

* update kv-cache concat method description

* quant-qwen leverage concatKV; add 8_0 to example main

* format 8_0 load

* remove trailing ,

* trailing line

* removed unnecessary contiguous calls

* Update candle-nn/src/kv_cache.rs

remove verbose kv-cache description

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

* Update candle-nn/src/kv_cache.rs

remove verbose kv-cache description

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

* Update candle-nn/src/kv_cache.rs

remove verbose kv-cache description

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

* Update candle-nn/src/kv_cache.rs

consolidate tests

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

* Update candle-transformers/src/models/quantized_qwen3.rs

Large improvements for kv_cache append quantized tensors when in contiguous layout

* Update candle-nn/src/kv_cache.rs

Since always using contiguous

* Update candle-nn/src/kv_cache.rs

after contiguous

* Update candle-nn/src/kv_cache.rs

after contiguous

* Update candle-transformers/src/models/quantized_qwen3.rs

contiguous called inside append

* Update candle-transformers/src/models/quantized_qwen3.rs

improves some devices but doesn't hurt others

* make k and v continguous post repeat in qwen3

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

Commit:
- Improved performance regressions since wgpu v25.x

- Corrected errors for the `wgpu_debug` feature
- Improved performance measurement documentation and debug file information

EricLBuehler committed:
Add `cublas_handle` api, update safetensors (#3192)

* Add cublas_handle api, update safetensors

* Add more quantized apis

* Make .vscode a .gitignore

ivarflakstad committed:
Update CI (#3194)

* Update CI

* I have no clue what was going on with this maturin file, but I don't like it

* update cuda container options

* Add compute cap to cuda wf

* Fix rust toolchain call

* update cuda ci runner and bindgen_cuda

EricLBuehler committed:
Add initial support for imatrix quantization (#3193)

anonenity committed:
add clear kv cache to quantized qwen3 weights (#3189)

amritsingh183 committed:
fix typo preventing usage on mac (#3201)

TimmyOVO committed:
CUDA: Fix integer reductions by removing +/-INF initialization (#3200)

* fix(cuda): fix integer reduction initialization
Replace hardcoded INFINITY/-INFINITY values with type-safe template functions for reduction initialization.
Using floating-point infinity values with integer types causes undefined behavior and crashes on newer GPU architectures like Blackwell.
The new template specializations use appropriate numeric_limits values for integer types while preserving the original behavior for floating-point types.

* fix(cuda): replace limits import with cuda std equivalents

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

amritsingh183 committed:
fix for https://github.com/huggingface/candle/issues/3203 (#3204)

* make qwen3 vl config public

Commit:
- Improved performance measurement debug output

- Fixed Clippy warning related to black_box usage
- Updated build.rs: functions inside shader files now preserve their source order instead of using a hash-based ordering
- fixed errors in wgsl, the shaders now use the wgsl select function, instead of defining its own

Commit:
- fixed an error in the index_select shader, if u32::MAX is used for the indexes

- improved documentation
- added a Auto Generated Comment Block on the wasm-test files

KimHenrikOtte committed:
- removed commented-out code

- corrected documentation

KimHenrikOtte committed:
Merge branch 'main' into wgpu_cleanup

KimHenrikOtte committed:
- fixed the return type of ArgMax/ArgMin Operations

- added load_quantization

haricot committed:
Add lld installation and test steps for Linux (#3213)

EricLBuehler committed:
Add dummy dtypes (#3195)

* Add dummy i32/i16/f6e2m3/f6e3m2/f4/f8e8m0 dtypes

* Metal fixes

* Fix candle-onnx build

* Apply review comments

* Residual fixes

* Apply review comments

* Apply review comments

* Revert some things

* Free more space

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

EricLBuehler committed:
Add more misc. changes from candle fork (#3196)

* Merge with fork

Co-authored-by Guoqing Bao <topon@outlook.com>

* Update sdpa

* Fix flash attn bf16 case

* Metal fixes

* Add metal methods

* Add new_private_buffer

* Fix metal tests

* Format

* Apply review comments

* Update CI (#3194)

* Update CI

* I have no clue what was going on with this maturin file, but I don't like it

* update cuda container options

* Add compute cap to cuda wf

* Fix rust toolchain call

* update cuda ci runner and bindgen_cuda

* Add initial support for imatrix quantization (#3193)

* add clear kv cache to quantized qwen3 weights (#3189)

* Fix metal bug

* Apply review comments

* Fix merge

* Add lld installation and test steps for Linux (#3213)

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>
Co-authored-by: anonenity <anonentity.mail@gmail.com>
Co-authored-by: Nicolas PASCAL <344493+haricot@users.noreply.github.com>

KimHenrikOtte committed:
- added rotary_emb_i

- added wgpu for rope tests
- removed default features for wasm tests (as for some reason the wgpu feature was enabled even for candle-examples)
- allow creation of u8 array
- copy use u32 Dtype

KimHenrikOtte committed:
fix wasm-helper

slckl committed:
.gitignore: add .zed to ignored editor configs (#3218)

mayocream committed:
chore(dep): bump cudarc to 0.18.1 (#3219)

EricLBuehler committed:
Hotfix: Bump float8 to 0.5.0 (#3223)

ivarflakstad committed:
[Metal] Make fast math mode optional (#3205)

* Add ability to toggle fast math mode in metal. Chose how to apply based on os version.

* Move available macro and friends to utils

* Isolate #[allow(deprecated)] to the actually deprecated method

* doc

* Use objc2::available macro instead

ivarflakstad committed:
Update pyo3 (#3202)

* Initial pyo3 update

* pyo3 onnx update

ivarflakstad committed:
[Metal] unary and affine improvements (#3230)

* Update unary.metal

* update metal unary tests

* Remove metal tiled unary kernels (now automated)

* Optimize metal affine

* Optimize metal powf

* Optimize metal elu

KimHenrikOtte committed:
- fixed storage_from_slice impl

- fixed candle-nn without wgpu feature

ivarflakstad committed:
[Metal] binary improvements (#3231)

amritsingh183 committed:
fix(metal): add missing softcapping field to AttnParams struct (#3233)

EricLBuehler committed:
Format sdpa (#3235)

EricLBuehler committed:
Fix metal argmax (#3238)

ivarflakstad committed:
[Metal] further improve unary and binary (#3239)

* Improve performance of contiguous unary/binary kernels

* Improved strided binary performance. Especially when only one of the tensors are strided.

ivarflakstad committed:
[Metal] cast improvements (#3241)

ivarflakstad committed:
[Metal] Improve ternary further (#3242)

ivarflakstad committed:
Bump candle version to 0.9.2-alpha.2 (#3248)

KimHenrikOtte committed:
improved performance for quantized matrix

KimHenrikOtte committed:
fix wgpu example

michaelfeil committed:
add candle flash attention 3 copyright markers (#3256)

DrJesseGlass committed:
fix: replace deprecated cudarc memcpy methods (#3228)

Replace deprecated cudarc memory copy methods:
- memcpy_dtov → clone_dtoh
- memcpy_stod → clone_htod

This fixes 12 deprecation warnings when building with --features cuda.

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

guoqingbao committed:
Support Fused MoE & Qwen3 GGUF MoE models (#3221)

* Support Fused MoE & Qwen3 GGUF MoE models

* Typo and cargo clippy fix

* Clippy fix

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

salmanmkc committed:
Upgrade GitHub Actions for Node 24 compatibility (#3255)

Signed-off-by: Salman Muin Kayser Chishti <13schishti@gmail.com>

skeet70 committed:
Adds onnx ops to support debertav3/piiranha (#3260)

* Adds ops to support debertav3/piiranha

Adds operations:
- Add
- Or
- Tile
- LessOrEqual
- GreaterOrEqual

Adds helpers `to_scalar_flexible` and `to_vec0_flexible` that both
allow for more broad input definitions that fit what some onnx models export
and are still scalar/vec0 but not "true" versions of them because the
model didn't squeeze them.

* Run cargo fmt

SpenserCai committed:
Add bilinear interpolation support (upsample_bilinear2d) (#3237)

* bilinear2d support init

* add bilinear2d utils test

* fix cu

* Extended testing

* fixed fmt and clippy warning

ivarflakstad committed:
Fix remnant memcpy_stod call (#3267)

slckl committed:
Sort on cuda fails when tensor size exceeds 1024 (#3271)

* candle-core: add asort_big test demonstrating issues with sorting larger
tensors on cuda

* candle-core/candle-kernels: lift the limitation of 1024 for sorting on
cuda

* skip asort_big test on metal for now

---------

Co-authored-by: Ivar Flakstad <69173633+ivarflakstad@users.noreply.github.com>

zackangelo committed:
make candle ops public (#3226)

clocksmith committed:
fix(quantized_gemma3): auto-detect GGUF metadata prefix for gemma-embedding and other Gemma model variants (#3274)

donjuanplatinum committed:
add HuberLoss (#3252)

* add HuberLoss and add Loss Trait

* 1. remove the LaTeX comment in loss.rs
2. add huberloss test

* change the huberloss Loss trait into the same approach as the other loss functions in this file

* cargo fmt

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

DanikVitek committed:
feat!: Make `ug` dependency optional (#3268)

* feat!: Make `ug` dep optional

* fix(example/mnist-training): Run all epochs

* doc(`candle-ug`): Crate documentation

* fix: feature-gate the `ComputePipeline` import

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

SpenserCai committed:
Add Z-Image Text-to-Image Generation Support (#3261)

* init z-image

* fixed patchify, unpatchify and latent

* update z_image examples readme

* fixed clippy and rustfmt

* fixed z_image example readme links

* support sdpa and flash-attn in Z-Image and fixed sdpa clippy warning

* fix some readme

* Update candle-transformers/src/models/z_image/transformer.rs

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

* support --model in example

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

Elvis339 committed:
fix(candle-kernels): conditionally link stdc++ for non-MSVC targets (#3278)

jacobgorm committed:
replace cutlass submodule references with explicit build step (#3234)

* replace cutlass submodule references with explicit build step

* address review comment: use Box leak trick also in attn-v3

* address review comment:

use "git clone --depth 1" instead of sparse checkout

for compatiblity with older git versions

* correct version in candle-flash-attn-build/Cargo.toml

* add top-level candle-flash-attn-build crate

* rustfmt

---------

Co-authored-by: Jacob Gorm Hansen <jhansen@dropbox.com>
Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

KimHenrikOtte committed:
fixed benchmarks for wgpu

improved matmul_quantized

KimHenrikOtte committed:
fixed clippy warnings

FerrisMind committed:
Rename compute capability defines in CUDA kernels (#3275)

guoqingbao committed:
Fix MoE WMMA kernel on V100 (#3282)

ivarflakstad committed:
[Metal] improve normalization (#3283)

amritsingh183 committed:
Fix BF16 conv_transpose2d using wrong kernel on Metal (#3279)

* Changed CONVT1D_OP to CONVT2D_OP for conv_transpose2d_bf16

* removing the test: the cause of the bug is so apparent.

KimHenrikOtte committed:
Merge branch 'main' into wgpu_cleanup

# Conflicts:
#	Cargo.toml
#	candle-core/Cargo.toml
#	candle-core/benches/benchmarks/binary.rs
#	candle-core/benches/benchmarks/unary.rs
#	candle-core/src/error.rs
#	candle-core/tests/tensor_tests.rs
#	candle-flash-attn/cutlass
#	candle-nn/benches/benchmarks/layer_norm.rs
#	candle-nn/benches/benchmarks/mod.rs
#	candle-transformers/src/lib.rs

KimHenrikOtte committed:
fixed merge errors

Anri-Lombard committed:
Mamba2 implementation (#3264)

danielclough committed:
feat: paddleocr-vl model and example (#3273)

KimHenrikOtte committed:
fixed benchmarks

KimHenrikOtte committed:
format document

KimHenrikOtte committed:
refactor: extract wgpu backend cache into separate crate

staceymelville committed:
chore(dep): bump cudarc to 0.18.2 (#3293)

DrJesseGlass committed:
Add SmolLM3: Full and Quantized Implementation (#3180)

* quantized and full SmolLM3

* include chrono for prompt

* resolve pub consist and unused var

* formatted

* last spacing in format

* add credits

* chat template

* integrate new chat template for smollm3 example

* fmt and clippy

* improve documentation / correct chat API in doc

* skip compile on documentation

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

DrJesseGlass committed:
example: add quantized qwen3 wasm with SIMD optimizations (#3159)

* minimal example

* toyed with cache

* serve cli

* cleaned up index

* readme

* cleaner interface

* improved formatting

* serve format

* index header unsloth

* demonstrate reasoning and format prompt

* no delay on button disable

* thinking based prompt; boolean switch; handle thinking reentry UI

* remove clear memory profile

* quant-qwen3 wasm using chat template

* thinking and no thinking tweakable mid conversation

* add discussion

* refactored example to use chat template; removed depracated fns

* clean up inputs and fmt

* remove unused logging import

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

EricLBuehler committed:
Hotfix: Remove fastmath from candle-kernels (#3309)

fffonion committed:
feat: add quantized lfm2 model support (#3244)

* feat: add quantized lfm2 model support

* fmt

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

guoqingbao committed:
Support new arch of GLM4 GGUF models (#2992)

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

drbh committed:
feat: simplify metal reduce kernels and standardize on u32 indexing (#3285)

* feat: reduce prefers min dtype based on stride size

* fix: cargo format fix

* Add temporary large/small reduce benchmarks

* Improved / simplified strided indexing

* Begin untangling pow2 concept from indexer

* remove unused get_strided_index_u64

* Make indexer_t handle both cont and strided. indexer.last_dim is a constant

* Remove Pow2Meta

* Remove redundant contiguous_indexer

* Remove redundant strided index impl

* Remove pow2 from kernel call/signature

* Some size_t -> ulong changes

* Use indexer_t for all reduce based kernels

* contiguous indexer last_dim default is 0

* u16 indexing does not provide speedup, and is an unlikely use case

* Let indexer_t dictate indexing dtype

* Introduce finalize concept. Tidy up redundant fns/macros

* Tidying up

* Use store concept instead of finalize

* Simplify arg reduce macros

* Tidying up

* Add reduce kernels for large tensors. Use existing reduce macro to add implementations

Remove large arg reduce kernels as we currently only support writing uint arg reduce results.

* Remove u64 indexed reduce kernels (u32 should suffice). Also max block_dim is 1024, so removing 2048 case from reduce kernels

* Remove suffix from reduce macro

* Remove IDX from reduce macros

* Tidy up reduce kernel call code

* Explicit u32 in reduce call code. Remove IDX from arg reduce macros

* Remove small reduce benchmark

---------

Co-authored-by: Ivar Flakstad <69173633+ivarflakstad@users.noreply.github.com>

ivarflakstad committed:
rms/layer norm accumulate in f32 for improved precision (#3315)

alvarobartt committed:
Remove `test.onnx` (uploaded by mistake?) (#3316)

KimHenrikOtte committed:
- Fixed shaders with incorrect access patterns (noticed in WASM tests) (cmp and reduce)

- Fixed gather implementation when the index is U32::MAX
- Implemented the first version of upsample_bilinear2d (still contains some errors; not all tests pass)
- Implemented rope and rope_thd (but rope_thd still fails some tests)
- improved candle-wasm-tests build.rs

KimHenrikOtte committed:
fixed upsample_bilinear2d

SpenserCai committed:
Metal GEMM Dynamic Tile Selection and Batch Collapse Optimization (#3313)

* mlx gemm opt init

* fix bug

* update

* opt

* opt

* update more shape to matmul benchmark

* remove metal_matmul_benchmark

---------

Co-authored-by: ivarflakstad <69173633+ivarflakstad@users.noreply.github.com>

ivarflakstad committed:
Update deps (#3320)

* Update deps

* add imageproc text feature

* Fix compilation

---------

Co-authored-by: Eric Buehler <ericlbuehler@gmail.com>

KimHenrikOtte committed:
renamed wgpu-compute-engine to wgpu-compute-layer

krampenschiesser committed:
CUDA Tensor::toDevice copies directly via memcopy_dtod (#3312)

* direct transfer for cuda

* allocate some random int data

* implement dummy device, test  same device logic

* update to latest cudarc version

* add cuda and metal checks

* change dependencies to use newer cudarc version

* reduce test to check for different devices

* Fix deps

* Clippy, fix workflow

* Format

* Fix handling for cuda

---------

Co-authored-by: Eric Buehler <ericlbuehler@gmail.com>

KimHenrikOtte committed:
removed unused code, made internal methods and structs private

ivarflakstad committed:
[Cuda] Use upstream bindgen_cuda crate (#3328)

* Use upstream bindgen_cuda crate. Use separate builders for each steps. Remove some cargo build warning messages

* bump bindgen_cuda

* fix

* Fix transfer_cuda_to_device test

ivarflakstad committed:
Bump candle version to 0.9.2 (#3329)

ivarflakstad committed:
Add dep versioning for candle-flash-attn-build (#3330)

KimHenrikOtte committed:
- removed unused code, made internal methods and structs private

- fixed MutexGuard across await point clippy warnings for native build (but not solved for wasm)

KimHenrikOtte committed:
- fixed rotarx_emb_thd impl

KimHenrikOtte committed:
- only add pipeline in the command_buffer if it changed

- add copy to staging_buffer at the of flush_gpu_command

KimHenrikOtte committed:
fixed quantized impl when using batch > 1

KimHenrikOtte committed:
format document

KimHenrikOtte committed:
format document

KimHenrikOtte committed:
Merge remote-tracking branch 'origin/main' into wgpu_cleanup

# Conflicts:
#	Cargo.toml

KimHenrikOtte committed:
- fixed warning when wgpu is not active

- fixed ternary_op_wgpu test (allow u8 buffer creation)
- fix warning in candle-wasm-tests
- removed wgpu feature from candle-test example

KimHenrikOtte committed:
- added tests

- improved doc comments
- restructured public api

KimHenrikOtte committed:
normalized layout for binary and unary operations

KimHenrikOtte committed:
test immediates1

KimHenrikOtte committed:
use defines to allow smaller intermmediate bufers

KimHenrikOtte committed:
fix memory leak

KimHenrikOtte committed:
- improved runtime shader generation performance

- a ShaderLoader::load now returns a Cow(a shader loader might return a static shader string, or may generate a shader in place)

KimHenrikOtte committed:
fixed browser impl

KimHenrikOtte committed:
format document

EricLBuehler committed:
Bump float8 to 0.7.0, cudarc to 0.19.1 (#3360)

EricLBuehler committed:
Bump float8 to 0.7.0, cudarc to 0.19.1 (#3360) (#3361)

KimHenrikOtte committed:
- Fixed Clippy warnings

- Simplified quantized shader loading
- Fixed `MutexGuard` across async methods; on WASM, the locks will be dropped before the await point
- Added multi-thread test

KimHenrikOtte committed:
format document

KimHenrikOtte committed:
Merge branch 'main' into wgpu_cleanup

# Conflicts:
#	candle-core/Cargo.toml

KimHenrikOtte committed:
improved diff with branch main

guoqingbao committed:
Use cudaforge for kernel build (#3346)

* Use cudaforge for kernel build

* Fix clippy

* Update cudaforge to v0.1.2

* Fix build candle-examples

---------

Co-authored-by: Eric Buehler <ericlbuehler@gmail.com>

FerrisMind committed:
Add candle-video library for text-to-video generation in README.md (#3365)

asglover committed:
fix: disable bf16 WMMA kernels on pre-Ampere GPUs (compute cap < 80) (#3367)

bf16 WMMA fragment types (nv_bfloat16 with nvcuda::wmma) are only
supported on sm_80+ (Ampere and later). On older architectures like
sm_75 (Turing/T4), compiling these fragments produces "incomplete type"
errors.

moe_wmma_gguf.cu already had #ifndef NO_BF16_KERNEL guards but
moe_wmma.cu was missing them, and the build script never passed the
-DNO_BF16_KERNEL define.

This commit:
- Adds matching #ifndef NO_BF16_KERNEL guards to moe_wmma.cu
- Updates build.rs to detect compute capability via cudaforge and
  pass -DNO_BF16_KERNEL when building for GPUs with compute cap < 80

danielclough committed:
feat: rwkv_v7 model and examples (#3372)

Co-authored-by: danielclough <danielclough@users.noreply.github.com>

KimHenrikOtte committed:
Merge branch 'main' into wgpu_cleanup

KimHenrikOtte committed:
improved wgpu candle-book, removed tracing_mutex

KimHenrikOtte committed:
- Removed the `wgpu` feature from `candle-test`.

- It previously appeared to enable the `wgpu` feature across the entire workspace.
- format document

KimHenrikOtte committed:
- fixed quantized implementation for wasm

- added quantized tests to candle-wasm-tests

KimHenrikOtte committed:
Fix filename casing: cargo.toml -> Cargo.toml

KimHenrikOtte committed:
Implement quantize_onto, quantize_imatrix, and quantize_imatrix_onto.

Remove cpu_flash_test and slow tests for WASM.
